src/libraries/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -1125,78 +1138,108 @@ public override Stream EndGetRequestStream(IAsyncResult asyncResult)
-                {
-                    // The System.Net.Http APIs require HttpRequestMessage headers to be properly divided between the request headers
-                    // collection and the request content headers collection for all well-known header names.  And custom headers
-                    // are only allowed in the request headers collection and not in the request content headers collection.
-                    if (IsWellKnownContentHeader(headerName))
-                    {
-                        // Create empty content so that we can send the entity-body header.
-                        request.Content ??= new ByteArrayContent(Array.Empty<byte>());
+            _sendRequestMessage.Version = ProtocolVersion;
 
-                        request.Content.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                    else
-                    {
-                        request.Headers.TryAddWithoutValidation(headerName, _webHeaderCollection[headerName!]);
-                    }
-                }
+            _sendRequestMessage.Headers.TransferEncodingChunked = SendChunked;
 
-                request.Headers.TransferEncodingChunked = SendChunked;
+            if (KeepAlive)
+            {
+                _sendRequestMessage.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
+            }
+            else
+            {
+                _sendRequestMessage.Headers.ConnectionClose = true;
+            }
 
-                if (KeepAlive)
-                {
-                    request.Headers.Connection.Add(HttpKnownHeaderNames.KeepAlive);
-                }
-                else
-                {
-                    request.Headers.ConnectionClose = true;
-                }
+            _sendRequestMessage.Version = ProtocolVersion;
+            HttpCompletionOption completionOption = _allowReadStreamBuffering ?
+                            HttpCompletionOption.ResponseContentRead : HttpCompletionOption.ResponseHeadersRead;
+            // If we're not buffering, there is no way to open the connection and not send the request without async.
+            // So we should use Async, if we're not buffering.
+            _sendRequestTask = async || !AllowWriteStreamBuffering ?
+                _httpClient.SendAsync(_sendRequestMessage, completionOption, _sendRequestCts.Token) :
+                Task.FromResult(_httpClient.Send(_sendRequestMessage, completionOption, _sendRequestCts.Token));
+
+            return _sendRequestTask!;
+        }
 
-                if (_servicePoint?.Expect100Continue == true)
+        private async Task<WebResponse> HandleResponse(bool async)
+        {
+            try
+            {
+                _requestStream?.EndWriteOnStreamBuffer();
+            }
+            catch (ObjectDisposedException)
+            {
+                // If the request stream was disposed, don't do anything.
```

Comment:
`EndWriteOnStreamBuffer` doesn't look like it would ever throw.
We can drop the try/catch here

Reply:
requestStream can be disposed before we call, that's why I'm keeping it here.

Reply:
But even if it is disposed, `EndWriteOnStreamBuffer` won't throw

Reply:
Because, I forgot to put it. Good catch, let me take a look.

Reply:
These are internal only helpers, so if we're going to ignore the exception anyway we might as well not throw it (especially given that the expectation here is that this stream will already be disposed in well behaved apps)

Reply:
Yes, that's what I'm doing right now, I will push changes in a min, I needed to check `streamBuffer` implementation first, but disposing or ending to write 2 times is not harming us.